### PR TITLE
Issue 129

### LINKED ISSUE #129: Polyhedral Contract Syntax
With: 

```
from pacti.utils.string_contract import StrContract
from pacti.terms.polyhedra.loaders import string_to_polyhedra_contract
```

We can write contracts in a human-friendly way like this:

```
# PWR:
# Parameters:
# - s: start index of the timeline variables
# - tstart: scheduled task instance start time
# - duration: scheduled task instance duration
# - generation: rate of battery charge during the task instance
def PWR_contract(s: int, tstart: float, duration: float, generation: float) -> tuple[int, list[IoContract]]:
  e = s+1
  spec = StrContract(
    inputs = [
      f"t{s}",    # Scheduled start time
      f"soc{s}",  # initial battery SOC
      f"d{s}",    # initial data volume
      f"e{s}",    # initial trajectory error
      f"r{s}",    # initial relative distance
    ],
    outputs = [
      f"t{e}",    # Scheduled end time
      f"soc{e}",  # final battery SOC
      f"d{e}",    # final data volume
      f"e{e}",    # final trajectory error
      f"r{e}",    # final relative distance
    ],
    assumptions = [
      # Scheduled task instance start time
      # t{s} == tstart
      f"t{s} <= {tstart}",
      f"-t{s} <= -{tstart}",
    ],
    guarantees = [
      # Scheduled task instance end time
      # t{e} - t{s} = duration
      f"t{e} - t{s} <= {duration}",
      f"t{s} - t{e} <= -{duration}",
    
      # Battery SOC charge
      # soc{e} - soc{s} >= duration*generation 
      f"soc{e} - soc{s} >= {duration*generation}",

      # no change to data volume
      # d{e} = d{s}
      f"d{e} - d{s} <= 0",
      f"d{s} - d{e} <= 0",

      # no change to trajectory error
      # e{e} = e{s}
      f"e{e} - e{s} <= 0",
      f"e{s} - e{e} <= 0",

      # no change to relative distance
      # r{e} = r{s}
      f"r{e} - r{s} <= 0",
      f"r{s} - r{e} <= 0",
    ])
  return e, string_to_polyhedra_contract(spec)
```

This approach delegates parsing the string expressions to Sympy, which accepts expressions that are not Polyhedral.
The resulting error messages can be difficult for users to understand.

This issue is about defining the syntax of Polyhedra constraints for parsing and rendering sketched as follows:

1) `LHS <= RHS`

where:

- `RHS` is a numerical constant
- `LHS` is of the form: `\sigma c V` where `c` is a constant, `V` is a variable.

For convenience, the following forms would be supported by rewrite as conjunctions of (1) above:

2) `| LHS | <= RHS`

Rewrite as the conjunction of:

- `LHS <= RHS`
- `-(LHS) <= RHS`

3) `| LHS | = 0`

Rewrite as the conjunction of:

- `LHS <= 0`
- `-(LHS) <= 0`

4) `LHS = RHS`

Rewrite as the conjunction of:

- `LHS <= RHS`
- `- (LHS) <= -(RHS)`

In the above:

- `-(RHS)` means negating the sign of the `RHS` constant
    e.g. `-(4)` becomes `-4` and `-(-3)` becomes `3` and `-(+2)` becomes `-2`
- `-(LHS)` means negating the effective sign of each atomic `c V` term in the `LHS` expression 
    e.g., `-(x-y)` becomes `-x+y`

The parser would reject any string expression that does not match any of (1,2,3,4).

This ticket entails two parts:

- [x] Parser

  This involves updating `pacti.terms.polyhedra.loaders.string_to_polyhedra_contract` to implement Regex-based parsing for the forms (1,2,3,4) above instead of using `Sympy`.

- [x] Pretty-printer

  Mapping the `pacti.iocontract.IoContract` internal representation of a Polyhedra contract according to the forms (1,2,3,4) above with a new function: `pacti.terms.polyhedra.loaders.polyhedra_contract_to_string`

### PR DESCRIPTION
**Pull request by NicolasRouquette**
Fixes #129

### CODE REVIEW

**Comment by NicolasRouquette** [comment]
I realize that I forgot to add support for parentheses. Please don't merge yet.

**Comment by NicolasRouquette** [comment]
@iincer Do you have any ideas how to handle this type checking problem?

```
pdm run duty check-types
✗ Type-checking (1)
  > mypy --strict --allow-any-generics --implicit-reexport --allow-untyped-calls --config-file=config/mypy.ini src tests
  src/pacti/terms/polyhedra/grammar.py:597: error: Unsupported left operand type for << ("ParserElement")  [operator]
```

This problem involves `pyparsing.Forward` construct and the associated `<<=` operator here:

```python
paren_terms <<= pp.Group("(" + terms + ")").set_parse_action(_parse_paren_terms).set_name("paren_terms_contents")
```

Strangely, the type error refers to `<<` even though the source code uses `<<=`.
Changing to the following makes no difference in the type checker error message nor in the behavior.
According to the doc, `<<=` is recommended over `<<`.

```python
paren_terms <<= pp.Group("(" + terms + ")").set_parse_action(_parse_paren_terms).set_name("paren_terms_contents")
```

There is some done about this topic:

https://pyparsing-docs.readthedocs.io/en/latest/HowToUsePyparsing.html#expression-operators
https://pyparsing-docs.readthedocs.io/en/latest/HowToUsePyparsing.html#special-subclasses

There are several examples of usage of this construct; see for example these search results:

https://github.com/search?q=repo%3Apyparsing%2Fpyparsing+%3C%3C%3D&type=code&p=2

Alas, it is unclear whether this project is performing type checking though there are several type annotations in the source code.

Is this `<<=` operator corresponding to the `__lshift__` method here: https://github.com/pyparsing/pyparsing/blob/master/pyparsing/core.py#L5395

**Comment by iincer** [comment]
Hi @NicolasRouquette, I made an update to this branch. The problem was that the function `pp.Forward().set_name()` in
```py
paren_terms = pp.Forward().set_name("paren_terms")
```
returns a type which is a parent of `pp.Forward`, and this parent class does not support the operator in question--the operator is defined for `pp.Forward`. The solution was to [cast the type to a subtype](https://mypy.readthedocs.io/en/stable/type_narrowing.html?highlight=cast#casts).

**Comment by NicolasRouquette** [comment]
@iincer This PR is now ready for review.

File: pyproject.toml [hunk 1]
```diff
@@ -20,6 +20,8 @@ dependencies = [
     "numpy>=1.24.1",
     "matplotlib>=3.6.2",
     "typing-extensions>=4.5.0",
+    "ipykernel>=6.22.0",
```
Comment (iincer) [inline pyproject.toml:23]: Does Pacti really need `ipykernel`?

**Review by iincer — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 1 above)
Comment (NicolasRouquette) [reply pyproject.toml:23]: Not it doesn't; I had used it when trying examples using notebooks. I'll remove it.

**Comment by NicolasRouquette** [comment]
Review notes:
- Relax the grammar for absolute terms to allow negative coefficients.
- For `leq` and `geq` expressions, add a check in the serializer to verify that all absolute term coefficients are positive after refactoring into `a <= <constant>` form.
- Make sure the `eql` expression type checks (the sides are `terms`)

**Comment by NicolasRouquette** [comment]
Review comments:

- add `polyhedra.syntax` namespace
- move `grammar.py` to `polyhedra.syntax`
- move the grammar dataclasses to `polyhedra.syntax.data.py`
- add unit tests that intentionally raise the polyhedra errors

**Review by iincer — APPROVED** [review]

**Pull request merged**